cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1076 @@
+
+    raft::update_host(
+      h_type_offsets.data(), type_offsets.data(), type_offsets.size(), handle.get_stream());
+
+    for (size_t idx = 0; idx < h_unique_types.size(); idx++) {
+      auto typ                   = h_unique_types[idx];
+      auto nr_elements_to_insert = (h_type_offsets[idx + 1] - h_type_offsets[idx]);
+
+      auto values_to_insert =
+        cugraph::allocate_dataframe_buffer<value_t>(nr_elements_to_insert, handle.get_stream());
+
+      auto zip_itr = thrust::make_zip_iterator(
+        thrust::make_tuple(edgelist_majors.begin(), edgelist_minors.begin()));
+
+      // FIXME: Can we avoid copying
+      thrust::copy(handle.get_thrust_policy(),
+                   zip_itr + h_type_offsets[idx],
+                   zip_itr + h_type_offsets[idx] + nr_elements_to_insert,
+                   cugraph::get_dataframe_buffer_begin(values_to_insert));
+
+      static_assert(std::is_same_v<
+                    typename thrust::iterator_traits<decltype(cugraph::get_dataframe_buffer_begin(
+                      values_to_insert))>::value_type,
+                    value_t>);
+
+      search_container.insert(
+        handle,
+        typ,
+        raft::device_span<edge_t>((*edgelist_ids).begin() + h_type_offsets[idx],
+                                  nr_elements_to_insert),
+        std::move(values_to_insert));
+    }
+  }
+
+  return search_container;
+}
+
+template <typename vertex_t,
+          typename edge_id_t,
+          typename edge_type_t,
+          typename EdgeTypeAndIdToSrcDstLookupContainerType,
+          bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+lookup_src_dst_from_edge_id_and_type(
+  raft::handle_t const& handle,
+  EdgeTypeAndIdToSrcDstLookupContainerType const& search_container,
+  raft::device_span<edge_id_t const> edge_ids_to_lookup,
+  edge_type_t edge_type_to_lookup)
+{
+  using value_t = typename EdgeTypeAndIdToSrcDstLookupContainerType::value_type;
+  static_assert(std::is_integral_v<edge_id_t>);
+  static_assert(std::is_integral_v<edge_type_t>);
+  static_assert(std::is_same_v<value_t, thrust::tuple<vertex_t, vertex_t>>);
+
+  static_assert(
+    std::is_same_v<typename EdgeTypeAndIdToSrcDstLookupContainerType::edge_id_type, edge_id_t>,
+    "edge_id_t must match EdgeTypeAndIdToSrcDstLookupContainerType::edge_id_type");
+  static_assert(
+    std::is_same_v<typename EdgeTypeAndIdToSrcDstLookupContainerType::edge_type_type, edge_type_t>,
+    "edge_type_t must match EdgeTypeAndIdToSrcDstLookupContainerType::edge_type_type ");
+
+  rmm::device_uvector<vertex_t> output_srcs(edge_ids_to_lookup.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> output_dsts(edge_ids_to_lookup.size(), handle.get_stream());
+
+  auto optional_value_buffer = search_container.src_dst_from_edge_id_and_type(
+    handle, edge_ids_to_lookup, edge_type_to_lookup, multi_gpu);
+
+  thrust::copy(
```

Comment:
Do we need a copy here?  Isn't the dataframe_buffer just a tuple of device vectors.  Seems like you could save the copy and just do something like `return std::move(optional_value_buffer);` or `return std::make_tuple(std::move(std::get<0>(optional_value_buffer)), std::move(std::get<1>(optional_value_buffer)));`?

@seunghwak ?